### PR TITLE
chunk, prod, pss, storage, swarm: trigger trojan chunk sending

### PR DESCRIPTION
**Pull request by santicomp2014**
**EXPERIMENTAL** 
(will not merge to `master`)

---
WIP for triggering trojan chunk sending used by implementation for global pinning.

this is invoked after trying to download content and failing, this will be a trigger to start the global pinning lifecycle.

Related issue: #2161

This PR:
- creates a new prod (recovery package) package for global pinning
- implements a way to trigger the sending of the trojan chunk; this should happen after failing to download content which the node understands is globally pinned.
- implemented by a fallback recoveryHook in netstore (similar to deliveryHook)
- should use feed-base retrieval but this maybe branch off into a separate PR.
  - sending should be done first with the hooks with a const target list hardwired in recovery package first
 - then it could be replaced with feed-based implementation

Remove `TODO`s for deprecated code after iteration

### CODE REVIEW

File: storage/netstore.go [hunk 1]
```diff
@@ -99,6 +100,7 @@ type NetStore struct {
 	requestGroup singleflight.Group
 	RemoteGet    RemoteGetFunc
 	logger       log.Logger
+	prod         *prod.Prod
```
Comment (zelig) [inline storage/netstore.go]: No, the dependency is the  other way round more natural.
Here  i would   assign a function or if you insist, an interface with      
`Recover` method

Comment (santicomp2014) [reply storage/netstore.go]: this was changed to be similar to Deliver hook.
This was one of the reasons there was an import cycle error.

**Review by zelig — APPROVED** [review]
i think this is done no?

File: storage/netstore.go [hunk 2]
```diff
@@ -193,6 +200,16 @@ func (n *NetStore) Get(ctx context.Context, mode chunk.ModeGet, req *Request) (c
...
 				if err != nil {
+					if n.recoveryCallback != nil {
+						n.recoveryCallback(ctx, ref)
+						time.Sleep(500 * time.Millisecond) // TODO: view what the ideal timeout is
+						ch, err = n.RemoteFetch(ctx, req, fi)
+						if err != nil {
+							// timeout and not found
+							return nil, err
+						}
+						return ch, nil
```
Comment (mortelli) [inline storage/netstore.go]:
```go
return n.RemoteFech(ctx, req, fi)
```

File: storage/netstore.go [hunk 3]
```diff
@@ -193,6 +200,16 @@ func (n *NetStore) Get(ctx context.Context, mode chunk.ModeGet, req *Request) (c
...
+					}
```
Comment (mortelli) [inline storage/netstore.go]: it was my understanding that we would wait and retry a certain amount of times after giving up on getting the chunk.

this only will call the hook, wait 500 ms, and try again, right? if it fails then, we're done.

we can defer this to a another issue for later if you wish, although maybe if we make use of the monitor at some point it's not so important to deal with right now.

what do you think?

File: prod/prod_test.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+			}
+		// TODO: change the timeout
+		case <-time.After(timeouts.FetcherGlobalTimeout):
+			t.Fatalf("no handler was called")
+		}
+	}
+
+}
+
+func GenerateTestRandomChunk() chunk.Chunk {
```
Comment (mortelli) [inline prod/prod_test.go]: may i ask why you are duplicating this function? just import it from its original package

File: prod/prod_test.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+	recoverFunc(ctx, chunk.ZeroAddr)
+
+	// verify the hook has been called
+	if handlerWasCalled != true {
+		t.Fatalf("unexpected result for prod Recover func, expected test variable to have a value of %v but is %v instead", true, handlerWasCalled)
+	}
+
+}
+
+func newMockLocalStore(t *testing.T, tags *chunk.Tags) *localstore.DB {
```
Comment (mortelli) [inline prod/prod_test.go]: this is also an exact duplicate of `pss.NewMockLocalStore`.

let's define it in 1 place and import it in the other.

if exported, please also add a comment briefly describing its purpose

File: prod/prod_test.go [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+	handlerWasCalled := false // test variable to check handler funcs are correctly retrieved
+
+	// register first handler
+	testHandler := func(ctx context.Context, targets [][]byte, topic trojan.Topic, payload []byte) (*pss.Monitor, error) {
+		handlerWasCalled = true
+		return nil, nil
+	}
+
+	recoverFunc := NewRecoveryHook(testHandler)
+	// call recoverFunc
```
Comment (mortelli) [inline prod/prod_test.go]: i think this comment is unnecessary

File: prod/prod_test.go [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+	localStore, err := localstore.New(dir, baseKey, &localstore.Options{Tags: tags})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	return localStore
+}
+
+// TestSenderCall verifies that pss send is being called correctly
```
Comment (mortelli) [inline prod/prod_test.go]: maybe i am missing something here, but i do not see `pss.Send` being used anywhere in this test

File: prod/prod_test.go [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+	handlerWasCalled := false // test variable to check handler funcs are correctly retrieved
+
+	// setup recovery hook
+	testHandler := func(ctx context.Context, targets [][]byte, topic trojan.Topic, payload []byte) (*pss.Monitor, error) {
+		handlerWasCalled = true
+		return nil, nil
+	}
+
+	//func(ctx context.Context, chunkAddress chunk.Address)
```
Comment (mortelli) [inline prod/prod_test.go]: this can be removed

File: prod/prod_test.go [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+	_, err := rand.Read(baseKey)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	baseAddress := network.NewBzzAddr(baseKey, baseKey)
+	// setup netstore
+	netStore := storage.NewNetStore(lstore, baseAddress)
+
+	handlerWasCalled := false // test variable to check handler funcs are correctly retrieved
```
Comment (mortelli) [inline prod/prod_test.go]: replace `handler` terminology with `hook` in this test as well

**Review by mortelli — CHANGES_REQUESTED** [review]
very nice job on the `TestSenderCall` func :clap: 

i think this needs just one more iteration and we're good to go :+1:

File: prod/prod.go [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+func NewRecoveryHook(send sender) RecoveryHook {
+	return func(ctx context.Context, chunkAddress chunk.Address) error {
+		targets, err := getPinners(chunkAddress)
+		if err != nil {
+			return err
+		}
+		payload := chunkAddress
+		topic := trojan.NewTopic("RECOVERY")
+
+		if _, err := send(ctx, targets, topic, payload); err != nil {
```
[the remark this answers is not in the document]
Comment (mortelli) [reply prod/prod.go:44]: see #2171

**Review by santicomp2014 — COMMENTED** [review]
I addressed all the comments.
Thanks for the feedback

File: prod/prod_test.go [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+
+	hookWasCalled := false // test variable to check hook func are correctly retrieved
+
+	// setup the hook
+	testHook := func(ctx context.Context, targets [][]byte, topic trojan.Topic, payload []byte) (*pss.Monitor, error) {
+		hookWasCalled = true
+		return nil, nil
+	}
+
+	// setup recovery hook with testHandler
```
Comment (mortelli) [inline prod/prod_test.go]: `testHandler` should be `testHook`

File: prod/prod_test.go (hunk 7 above)
Comment (mortelli) [reply prod/prod_test.go]: much better, thanks

**Review by mortelli — APPROVED** [review]
good stuff! thank you for addressing everything.

File: prod/prod_test.go (hunk 11 above)
Comment (santicomp2014) [reply prod/prod_test.go]: Missed that one.
Fixed

**Pull request merged**